File: game.py
import pygame

pygame.init()

WIDTH, HEIGHT = 940, 940
BOARD_SIZE = 280

WHITE = (255, 255, 255)
BLACK = (0, 0, 0)
RED = (200, 0, 0)
DARK_RED = (50, 0, 0)
DARK_BLUE = (0, 0, 50)

img_X = pygame.image.load("assets/x.png")
img_X = pygame.transform.scale_by(img_X, 4)
big_img_X = pygame.transform.scale_by(img_X, 5)
img_O = pygame.image.load("assets/o.png")
img_O = pygame.transform.scale_by(img_O, 4)
big_img_O = pygame.transform.scale_by(img_O, 5)
imgs = {1: img_X, 0: img_O}
big_imgs = {1: big_img_X, 0: big_img_O}

screen = pygame.display.set_mode((WIDTH, HEIGHT))
big_font = pygame.font.SysFont("Helvetica", 54)
small_font = pygame.font.SysFont("Helvetica", 28)
pygame.display.set_caption("Ultimate Tic-Tac-Toe")


class Board:
    def __init__(self, pos):
        self.grid = [None] * 9
        self.pos = pos

        self.scored = None

        self.rect = pygame.rect.Rect(pos[0], pos[1], BOARD_SIZE, BOARD_SIZE)

        self.click_rects = []
        for n in range(9):
            self.click_rects.append(
                pygame.rect.Rect(pos[0] + 20 + (82 * (n % 3)), pos[1] + 20 + (82 * (n // 3)), 76, 76))

    def check(self):
        # check for vertical scores
        for i in range(3):
            if self.grid[i] == self.grid[i + 3] == self.grid[i + 6] and self.grid[i] in [0, 1]:
                self.scored = self.grid[i]
        # check for horizontal scores
        for i in range(3):
            if self.grid[i * 3] == self.grid[i * 3 + 1] == self.grid[i * 3 + 2] and self.grid[i * 3] in [0, 1]:
                self.scored = self.grid[i * 3]
        # top left to bottom right diagonal
        if self.grid[0] == self.grid[4] == self.grid[8] and self.grid[0] in [0, 1]:
            self.scored = self.grid[0]
        # top right to bottom left diagonal
        if self.grid[2] == self.grid[4] == self.grid[6] and self.grid[2] in [0, 1]:
            self.scored = self.grid[2]

        if None not in self.grid and self.scored is None:
            self.scored = 2

        if self.scored is not None:
            big_check()

    def mark(self, n):
        global turn, allowed
        if self.grid[n] is not None:
            return

        self.grid[n] = turn
        turn = 0 if turn == 1 else 1
        self.check()

        # calculate new allowed board
        if DEBUG:
            allowed = None
        else:
            allowed = boards[n]
            if allowed.scored is not None:
                allowed = None

    def draw(self):
        """
        we have 240, because we leave a 20 margin

        the actual bounds are 20-260x20-260


        :param pos:
        :return:
        """
        temp_surface = pygame.Surface((BOARD_SIZE, BOARD_SIZE), pygame.SRCALPHA)
        pos = self.pos
        rect1 = (96, 20, 6, 240)
        rect2 = (178, 20, 6, 240)
        pygame.draw.rect(temp_surface, WHITE, rect1)
        pygame.draw.rect(temp_surface, WHITE, rect2)

        rect1 = (20, 96, 240, 6)
        rect2 = (20, 178, 240, 6)
        pygame.draw.rect(temp_surface, WHITE, rect1)
        pygame.draw.rect(temp_surface, WHITE, rect2)

        if self.scored is not None:
            temp_surface.set_alpha(128)
            if self.scored in [0, 1]:
                screen.blit(big_imgs[self.scored], (30 + pos[0], 30 + pos[1]))

        for n, i in enumerate(self.grid):
            if i is not None:
                new_pos = (36 + (82 * (n % 3)), 36 + (82 * (n // 3)))
                temp_surface.blit(imgs[i], new_pos)

        return temp_surface


def big_check():
    global won, allowed
    for i in range(3):
        if boards[i].scored == boards[i + 3].scored == boards[i + 6].scored and boards[i].scored in [0, 1]:
            won = boards[i].scored
        # check for horizontal scores
    for i in range(3):
        if boards[i * 3].scored == boards[i * 3 + 1].scored == boards[i * 3 + 2].scored and boards[
            i * 3].scored in [0, 1]:
            won = boards[i * 3].scored
        # top left to bottom right diagonal
    if boards[0].scored == boards[4].scored == boards[8].scored and boards[0].scored in [0, 1]:
        won = boards[0].scored
        # top right to bottom left diagonal
    if boards[2].scored == boards[4].scored == boards[6].scored and boards[2].scored in [0, 1]:
        won = boards[2].scored
    if won is not None:
        allowed = None


def draw_big_board():
    rect1 = (310, 20, 20, 900)
    rect2 = (610, 20, 20, 900)
    pygame.draw.rect(screen, WHITE, rect1)
    pygame.draw.rect(screen, WHITE, rect2)

    rect1 = (20, 310, 900, 20)
    rect2 = (20, 610, 900, 20)
    pygame.draw.rect(screen, WHITE, rect1)
    pygame.draw.rect(screen, WHITE, rect2)


def render():
    if won is not None and render_won_screen:
        screen.fill(BLACK)
        text1 = big_font.render(f"{'X' if won else 'O'} won the game!", False, WHITE)
        screen.blit(text1, (470 - text1.get_width() // 2, 400))
        subtext = small_font.render("Press v to view board, r to restart", False, WHITE)
        screen.blit(subtext, (470 - subtext.get_width() // 2, 500))
        return

    screen.fill(DARK_RED if turn == 1 else DARK_BLUE)

    draw_big_board()

    # draw individual boards
    for board in boards:
        temp_surface = board.draw()
        if allowed:
            if board is not allowed:
                temp_surface.set_alpha(128)

        screen.blit(temp_surface, board.pos)


def initialize_boards():
    for n in range(9):
        pos = (30 + (300 * (n % 3)), 30 + (300 * (n // 3)))
        boards.append(Board(pos))


boards = []
initialize_boards()
turn = 1
won = None
render_won_screen = True

allowed = None
DEBUG = False

running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_d:
                DEBUG = not DEBUG
            if event.key == pygame.K_v:
                render_won_screen = not render_won_screen
            if event.key == pygame.K_r:
                boards = []
                won = None
                render_won_screen = True
                turn = 1
                allowed = None
                initialize_boards()

        if event.type == pygame.MOUSEBUTTONDOWN:
            if won is not None:
                continue
            pos = pygame.mouse.get_pos()
            for board_n, board in enumerate(boards):
                if board.rect.collidepoint(pos):
                    break
            # check if this is the board we're allowed to play in
            if allowed:
                if board != allowed:
                    continue
            # check if board is already scored
            if board.scored is not None:
                continue
            for cell, cell_rect in enumerate(board.click_rects):
                if cell_rect.collidepoint(pos):
                    board.mark(cell)

    render()
    pygame.display.flip()
pygame.quit()

File: game.js
const canvas = document.getElementById('gameCanvas');
const ctx = canvas.getContext('2d');
const statusDisplay = document.getElementById('status');
const restartButton = document.getElementById('restart');
const viewBoardButton = document.getElementById('viewBoard');
// TODO - add undo button, maybe fix spacing if i care enough
canvas.height = window.innerHeight - 10;
canvas.width = canvas.height;
viewBoardButton.hidden = true;

// COLORS
const WHITE = '#FFFFFF';
const BLACK = '#000000';
const RED = '#FF0000';
const TRANSL_RED = 'rgba(255, 0, 0, 0.8)'
const BLUE = '#0000FF';
const TRANSL_BLUE = 'rgba(0, 0, 255, 0.8)'
const DARK_RED = '#320000';
const DARK_BLUE = '#000032';
const GRAY = '#888888';

// RESIZING STUFF
let outerMargin, innerMargin;
let bigBarWidth, smallBarWidth;
let pos;
let availSpace = canvas.width - outerMargin*2;
let boardSize, boardReal, cellSize;
let imgMargin, imgSize;
let bigFont, lilFont;

// TEMPs
let relX;
let relY;
let notFull = 0;

// GAME VARS
let turn = 1;
let allowed = null;
let boards = [];
let won = null;
let gameOver = null;
let debug_mode = false; // developer only so no touchies pls - it lets you play anywhere

// DRAWING
imgO = new Image()
imgO.src = "assets/o.png"
imgX = new Image()
imgX.src = "assets/x.png"

// GRAPHICAL FUNCTIONS
function handleResize () {
    canvas.height = window.innerHeight - 10;
    canvas.width = canvas.height;

    bigBarWidth = Math.ceil(canvas.height / 60); // 14 px on my sscreen size
    smallBarWidth = Math.ceil(canvas.height / 200); // 4 px on my sscreen size

    outerMargin = Math.floor(canvas.height / 40) // 20 px on my screen size
    innerMargin = Math.floor(canvas.height / 40) // 20 px on my sscreen size

    availSpace = canvas.width - outerMargin*2
    boardSize = Math.floor((availSpace - bigBarWidth*2) / 3)
    boardReal = boardSize - innerMargin * 2

    cellSize = Math.floor((boardReal - smallBarWidth * 2) / 3);

    imgMargin = Math.floor(cellSize/10);
    imgSize = cellSize - 2*imgMargin;

    bigFont = Math.floor(canvas.width / 30)
    lilFont = Math.floor(canvas.width / 45)

    setBoardsPos();
    drawGame();
}

function setBoardsPos() {
    for (let i=0; i<9; i++) {
        pos = [outerMargin+innerMargin, outerMargin+innerMargin]
        pos[0] += (boardSize+bigBarWidth)*(i % 3)
        pos[1] += (boardSize+bigBarWidth)*Math.floor(i / 3)
        boards[i].tl = pos;
        boards[i].br = [pos[0] + boardReal, pos[1] + boardReal];
    }
}

// LOCATION FUNCTIONS 
function getMouseBoard(x, y) {
    for (let i=0; i<9; i++) {
        if (x >= boards[i].tl[0] && x <= boards[i].br[0] &&
            y >= boards[i].tl[1] && y <= boards[i].br[1]) {
                return i;
        }
    }
}

function getMouseRelCell(x, y) {
    for (let i=0; i<9; i++) {
        // do frogs bounce
        cellX = (i % 3) * (cellSize + smallBarWidth);
        cellY = Math.floor(i / 3) * (cellSize + smallBarWidth);
        if (x >= cellX && x <= cellX + cellSize &&
            y >= cellY && y <= cellY + cellSize) {
                return i;
        }
    }
}

// GAME DEFINITIONS
class Board {
    constructor(pos) {
        this.grid = [null, null, null, null, null, null, null, null, null];
        this.scored = null;
        this.tl = pos;
        this.br = [pos[0] + boardReal, pos[1] + boardReal];
        
    }

    check() {
        notFull = 0;
        // check to see if grid is filled
        for (let i=0; i<9; i++) {
            if (this.grid[i] !== null)
                notFull += 1;
        }
        if (notFull === 9){
            this.scored = 2;
            return
        }

        for (let i=0; i<3; i++) {
            // check horizontal
            if (this.grid[i*3] === this.grid[i*3 + 1] 
                && this.grid[i*3] === this.grid[i*3 + 2] 
                && this.grid[i*3] !== null)
                this.scored = this.grid[i*3];
            // vertical
            if (this.grid[i] === this.grid[i+3]
                && this.grid[i] === this.grid[i+6] 
                && this.grid[i] !== null) {this.scored = this.grid[i];}
        }
        
        // diagonal tl-br
        if (this.grid[0] === this.grid[4] 
            && this.grid[0] === this.grid[8] 
            && this.grid[0] !== null)
            this.scored = this.grid[0];
        // diagonal tr-bl
        if (this.grid[2] === this.grid[4]
            && this.grid[2] === this.grid[6] 
            && this.grid[2] !== null)
            this.scored = this.grid[2];
        
        
        if (this.scored !== null)
            bigCheck();
    }

    mark(cellNum) {
        this.grid[cellNum] = turn;
    }

    draw(dulled) {
        ctx.imageSmoothingEnabled = false;
        if (this.scored === 1 || this.scored === 0) {
            
            ctx.drawImage(this.scored === 1 ? imgX : imgO, this.tl[0]+imgMargin, this.tl[1]+imgMargin, boardReal-2*imgMargin, boardReal-2*imgMargin)
            return
        }
        ctx.fillStyle = dulled === false ? WHITE : GRAY;

        // vertical
        ctx.fillRect(this.tl[0]+cellSize, this.tl[1], smallBarWidth, boardReal);
        ctx.fillRect(this.tl[0]+boardReal-cellSize-smallBarWidth, this.tl[1], smallBarWidth, boardReal);

        // horizontal
        ctx.fillRect(this.tl[0], this.tl[1]+cellSize, boardReal, smallBarWidth);
        ctx.fillRect(this.tl[0], this.tl[1]+boardReal-cellSize-smallBarWidth, boardReal, smallBarWidth);
        
        // draw X/Os 
        for (let i = 0; i<9; i++) {
            if (this.grid[i] == null)
                continue
            ctx.fillStyle = this.grid[i] === 1 ? RED : BLUE
            relX = this.tl[0] + (cellSize+smallBarWidth)*(i % 3)
            relY = this.tl[1] + (cellSize+smallBarWidth)*Math.floor(i/3)
            
            ctx.drawImage(this.grid[i] === 1 ? imgX : imgO, relX+imgMargin, relY+imgMargin, imgSize, imgSize)
            //ctx.fillRect(relX, relY, cellSize, cellSize)

            // TODO - add image loading -------------------------
        }
    }
};

function initBoards() {
    for (let i=0; i<9; i++) {
        pos = [outerMargin+innerMargin, outerMargin+innerMargin]
        pos[0] += (boardSize+bigBarWidth)*(i % 3)
        pos[1] += (boardSize+bigBarWidth)*Math.floor(i / 3)
        boards.push(new Board(pos))
    }
}

initBoards();
// ok lets be real whos reading the code for my crappy web game???
// i appreciate you whoever you are
// youre awesome

function bigCheck() {
    notFull = 0;
    // check to see if grid is filled
    for (let i=0; i<9; i++) {
        if (boards[i].scored !== null)
            notFull += 1;
    }
    if (notFull === 9){
        won = 2;
        return
    }
    // check horizontal scores
    for (let i=0; i<3; i++) {
        if (boards[i*3].scored === boards[i*3 + 1].scored 
            && boards[i*3].scored === boards[i*3 + 2].scored 
            && (0 === boards[i*3].scored
            || 1 === boards[i*3].scored))
            won = boards[i*3].scored;
        if (boards[i].scored === boards[i+3].scored 
            && boards[i].scored === boards[i+6].scored 
            && (0 === boards[i].scored
            || 1 === boards[i].scored))
            won = boards[i].scored;

    }
    // diagonal tl-br
    if (boards[0].scored === boards[4].scored
        && boards[0].scored === boards[8].scored 
        && (0 === boards[0].scored
        || 1 === boards[0].scored))
        won = boards[0].scored;
    // diagonal tr-bl
    if (boards[2].scored === boards[4].scored
        && boards[2].scored === boards[6].scored 
        && (0 === boards[2].scored
        || 1 === boards[2].scored))
        won = boards[2].scored;
    
    if (won !== null) {
        gameOver = 1;
        viewBoardButton.hidden = false;
    }
}

handleResize();

function handleClick(e) {
    const rect = canvas.getBoundingClientRect();
    const x = e.clientX - rect.left;
    const y = e.clientY - rect.top;
    // on click, we should:
    // check that the game isn't over already
    if (gameOver !== null)
        return
    // get board that's being clicked on
    boardNum = getMouseBoard(x, y);
    // validate board existence
    if (boardNum === undefined)
        return;
    // validate board allowedness
    if (allowed !== null && debug_mode === false) {
        if (boardNum !== allowed) 
            return;
    }
    // get the cell that's being clicked on
    relX = x - boards[boardNum].tl[0];
    relY = y - boards[boardNum].tl[1];
    cellNum = getMouseRelCell(relX, relY);

    // if its populated, pass
    if (boards[boardNum].grid[cellNum] !== null) 
        return;

    // mark the cell
    boards[boardNum].mark(cellNum);

    // check the board
    boards[boardNum].check();

    // switch turns
    turn = turn === 1 ? 0 : 1;


    // define new allowed (unless in debug_mode)
    if (boards[cellNum].scored !== null) 
        allowed = null
    else if (debug_mode === false)
        allowed = cellNum

    drawGame();
}


function drawGame() {
    
    ctx.fillStyle = turn === 1 ? DARK_RED : DARK_BLUE;
    if (gameOver)
        ctx.fillStyle = "rgb(60, 60, 60)";
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    // 0: if game is won, show win overlay and return
    if (won !== null) {
        let userText;
        if (won == 2) {
            ctx.fillStyle = BLACK;
            userText = "Nobody"; 
        }
        else {
            ctx.fillStyle = won === 1 ? DARK_RED : DARK_BLUE;
            userText = won === 1 ? "X" : "O"
        }
        ctx.fillRect(0, 0, canvas.width, canvas.height);
        ctx.fillStyle = WHITE;
        ctx.font = String(bigFont) + "pt Helvetica"
        let mst = ctx.measureText(userText + " won the game! Well done.")
        ctx.fillText(userText + " won the game! Well done.", canvas.width/2 - mst.width/2, canvas.height/2)
        // no way ur still reading my code!!
        // awesomeeeeeee
        // i think im in love
        // (with you)
        // 
        // ...
        //
        // ...
        //
        // see that's funny because i dont know who you are and so im confessing my love to a complete stranger
        // 
        // but thats the trick - maybe i DO love you and i just dont know that ur reading my code
        // or maybe i think youre a really awful person and this is extra weird
        // i doubt the latter cus there aren't many people i think that about
        // either way ur probbaly cool so yea
        // thanks
        // for being cool
        // i guess
        ctx.font = String(lilFont) + "pt Helvetica"
        let mst2 = ctx.measureText("Use the sidebar buttons to continue.")
        ctx.fillText("Use the sidebar buttons to continue.", canvas.width/2 - mst2.width/2, canvas.height/2+(bigFont+lilFont)*0.8)
        return
    }

    // 1: draw the big board
    ctx.fillStyle = WHITE;
    // 1.1: verticals
    ctx.fillRect(outerMargin+boardSize, outerMargin, bigBarWidth, availSpace);
    ctx.fillRect(canvas.height - outerMargin - boardSize - bigBarWidth, outerMargin, bigBarWidth, availSpace);
    // 1.2: horizontals
    ctx.fillRect(outerMargin, outerMargin+boardSize, availSpace, bigBarWidth);
    ctx.fillRect(outerMargin, canvas.width - outerMargin - boardSize - bigBarWidth, availSpace, bigBarWidth);

    // 2: draw each small board
    for (let i=0; i<9; i++) {
        if (allowed !== null && allowed !== i)
            boards[i].draw(true);
        else
            boards[i].draw(false);
    }
}

canvas.addEventListener('click', handleClick);

window.addEventListener("resize", handleResize);

restartButton.addEventListener("click", (e) => {
    turn = 1;
    allowed = null;
    boards = [];
    initBoards();
    won = null;
    gameOver = null;
    viewBoardButton.hidden = true;
    // the potatos are taking over my miiiiiiiiiind
    drawGame();
});

viewBoardButton.addEventListener("click", (e) => {
    won = null;
    allowed = null;
    drawGame();
});

drawGame();